test/testblacklist.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ *    This software is in the public domain, furnished "as is", without technical
+ *    support, and with no warranty, express or implied, as to its usefulness for
+ *    any purpose.
+ *
+ */
+
+#include <QtTest>
+#include "syncenginetestutils.h"
+#include <syncengine.h>
+
+using namespace OCC;
+
+SyncFileItemPtr findItem(const QSignalSpy &spy, const QString &path)
+{
+    for (const QList<QVariant> &args : spy) {
+        auto item = args[0].value<SyncFileItemPtr>();
+        if (item->destination() == path)
+            return item;
+    }
+    return SyncFileItemPtr(new SyncFileItem);
+}
+
+SyncJournalFileRecord journalRecord(FakeFolder &folder, const QByteArray &path)
+{
+    SyncJournalFileRecord rec;
+    folder.syncJournal().getFileRecord(path, &rec);
+    return rec;
+}
+
+class TestBlacklist : public QObject
+{
+    Q_OBJECT
+
+private slots:
+    void testBlacklistBasic_data()
+    {
+        QTest::addColumn<bool>("remote");
+        QTest::newRow("remote") << true;
+        QTest::newRow("local") << false;
+    }
+
+    void testBlacklistBasic()
+    {
+        QFETCH(bool, remote);
+
+        FakeFolder fakeFolder{ FileInfo::A12_B12_C12_S12() };
+        QCOMPARE(fakeFolder.currentLocalState(), fakeFolder.currentRemoteState());
+        QSignalSpy completeSpy(&fakeFolder.syncEngine(), SIGNAL(itemCompleted(const SyncFileItemPtr &)));
+
+        auto &modifier = remote ? fakeFolder.remoteModifier() : fakeFolder.localModifier();
+
+        auto cleanup = [&]() {
+            completeSpy.clear();
+        };
+
+        auto initialEtag = journalRecord(fakeFolder, "A")._etag;
+        QVERIFY(!initialEtag.isEmpty());
+
+        // The first sync and the download will fail - the item will be blacklisted
+        modifier.insert("A/new");
+        fakeFolder.serverErrorPaths().append("A/new", 500); // will be blacklisted
+        QVERIFY(!fakeFolder.syncOnce());
+        {
+            auto it = findItem(completeSpy, "A/new");
+            QVERIFY(it);
+            QCOMPARE(it->_status, SyncFileItem::NormalError); // initial error visible
+            QCOMPARE(it->_instruction, CSYNC_INSTRUCTION_NEW);
+
+            auto entry = fakeFolder.syncJournal().errorBlacklistEntry("A/new");
+            QVERIFY(entry.isValid());
+            QCOMPARE(entry._errorCategory, SyncJournalErrorBlacklistRecord::Normal);
+            QCOMPARE(entry._retryCount, 1);
+            QVERIFY(entry._ignoreDuration > 0);
+
+            if (remote)
+                QCOMPARE(journalRecord(fakeFolder, "A")._etag, initialEtag);
+        }
+        cleanup();
+
+        // Ignored during the second run - but soft errors are also errors
+        QVERIFY(!fakeFolder.syncOnce());
+        {
+            auto it = findItem(completeSpy, "A/new");
+            QVERIFY(it);
+            QCOMPARE(it->_status, SyncFileItem::BlacklistedError);
+            QCOMPARE(it->_instruction, CSYNC_INSTRUCTION_IGNORE); // no retry happened!
```

Comment:
Maybe there could be a counter of the amount of PUT calls in addition to that.

Reply:
Added.